buildSrc/src/main/java/org/elasticsearch/gradle/precommit/PrecommitPlugin.java

```diff
@@ -0,0 +1,59 @@
+/*
+ * Licensed to Elasticsearch under one or more contributor
+ * license agreements. See the NOTICE file distributed with
+ * this work for additional information regarding copyright
+ * ownership. Elasticsearch licenses this file to you under
+ * the Apache License, Version 2.0 (the "License"); you may
+ * not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing,
+ * software distributed under the License is distributed on an
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+ * KIND, either express or implied.  See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+
+package org.elasticsearch.gradle.precommit;
+
+import org.elasticsearch.gradle.util.GradleUtils;
+import org.gradle.api.DefaultTask;
+import org.gradle.api.Plugin;
+import org.gradle.api.Project;
+import org.gradle.api.Task;
+import org.gradle.api.artifacts.Dependency;
+import org.gradle.api.plugins.JavaBasePlugin;
+import org.gradle.api.plugins.JavaPlugin;
+import org.gradle.api.tasks.SourceSet;
+import org.gradle.api.tasks.TaskProvider;
+import org.gradle.language.base.plugins.LifecycleBasePlugin;
+
+/**
+ * Base plugin for adding a precommit task.
+ */
+public abstract class PrecommitPlugin implements Plugin<Project> {
```

Comment:
I think it would be better to not extend the other PrecommitPlugins not extend this Plugin but use composition here. So we would have a PreCommit plugin that is applied by all other specific PreCommitPlugins (e.g. the CheckstylePreCommitPlugin). The PreCommitPlugin would only create a 'precommit' task that depends on all tasks of a certain `marker` type like `tasks.withType(PreCommitTask.class)`. The other specific precommit plugins would only need to ensure that the tasks they create are of type PreCommitTask which can be just common interface.

Reply:
Thinking more about it an alternative would be if the specific PreCommit plugins are just responsible to wire things against the precommit task. That way you could skip the specific common task type. so these plugins would look always like

``` 
...
apply(PreCommitPlugin)
// register tasks I want to wire
tasks.named("precommit").configure { dependsOn 'specificTask; }
...
```

Reply:
> I think it would be better to not extend this Plugin in other PrecommitPlugins but use composition here

Using inheritance allows us ensure the correct wirings happen. With composition, we can forget to hook up the new task to precommit, but with inheritance we enforce this because the class and createTask methods are abstract. Additionally, it wouldn't make sense to have `precommit` alone, it only makes sense in the context of having a concrete task like forbiddenapis.

> The other specific precommit plugins would only need to ensure that the tasks they create are of type PreCommitTask

We can't guarantee this since some tasks are external, like forbidden apis or checkstyle.

Reply:
I see. I would still split out the generation of the precommit task to a small (internal) plugin (e.g. PrecommitBase Plugin. By internal I mean we would not need to provide a public id for the plugin or even make it package protected. This way the configuration is 

a) ensure this task registration is done in a gradle idiomatic way
b) we ensure its only applied by the specific PreCommit Plugins.

this is a common pattern in gradle where a plugin just provides a lifecycle task and other plugins hook into this lifecycle.

Reply:
I think a separate plugin that creates the precommit task might be the way to go as well. Primary reason being that `maybeRegister` does exception-based flow control, which is incredibly expensive and basically negates all benefits of lazy tasks. In fact, I'd recommend we completely remove that method so folks don't use it. We are throwing hundreds of exceptions here during configuration time. Filling all those stack frames has a big penalty.

Reply:
I don't know enough yet to understand how much this method is used. But for sure we could deprecate it.

Reply:
I believe this is the only usage. Since that is the case we should remove it and find another way here. The `findByName` method has the same problem. When investigating doing some more lazy task stuff I found that switching creation to registration leveraging `findByName` was _slower_ due to all the additional exceptions. We should try and get this added to Gradle core in a proper API instead.

Reply:
👍 small general thought on findByName. `findByName` often is an indicator of weak modelling that can be improved. I think in the build there's a lot of logic that is based on tasks that are around which should be more about what type of project (which plugins are applied) are we dealing with.

Reply:
>  findByName often is an indicator of weak modelling that can be improved
That's definitely true. If you have to check if a task exists that indicates a code smell.

Reply:
After chatting with Ryan I think I'm fine with plugin class extension. I'm less concerned about extending plugins when the superclass is abstract, that at least indicates this is purposeful. Extending non-abstract plugins is a no-no, and composition should be used instead as you say.

We can then have the abstract plugin apply a plugin which creates the `precommit` task that way Gradle handles idempotency for us rather than using `maybeRegister`.

Reply:
> We can then have the abstract plugin apply a plugin which creates the precommit task that way Gradle handles idempotency for us rather than using maybeRegister.

👍

Reply:
I pushed b5dff57 which no longer uses maybeCreate. Instead we have an inner plugin that creates the task, but we still use inheritance to ensure we don't forget to add a task to it.